File: src/services/smsoip.ts
/**
 * SMSoIP (ap2005) Service Handler
 *
 * Builds the ApplicationConfig for SMS over IP entitlements.
 */

import type { ApplicationConfig } from '../protocol/responseTypes.js';
import { EntitlementStatus } from '../protocol/statusCodes.js';

interface SmsOipConfigData {
  addresses?: Array<{ addrType: string; addr: string }>;
}

const DEFAULT_ADDRESSES = [
  { addrType: '1', addr: 'smsc.operator.com' },
];

export function buildSmsOipConfig(
  status: number,
  provStatus: number,
  tcStatus: number,
  configData?: unknown,
): ApplicationConfig {
  const config = (configData ?? {}) as SmsOipConfigData;

  const result: ApplicationConfig = {
    appId: 'ap2005',
    entitlementStatus: status,
    tcStatus,
    provStatus,
  };

  if (status === EntitlementStatus.ENABLED) {
    result.addrStatus = 1;
    result.addresses = config.addresses ?? DEFAULT_ADDRESSES;
  } else {
    result.addrStatus = 0;
  }

  return result;
}

File: internal/services/smsoip.go
package services

import (
	"github.com/hkassaei/poc-entitlements-as-a-service-go/internal/protocol"
)

// SmsOipConfigData holds configData for ap2005.
type SmsOipConfigData struct {
	Addresses []protocol.AddressConfig `json:"addresses,omitempty"`
}

var _defaultSmsOipAddresses = []protocol.AddressConfig{
	{AddrType: "1", Addr: "smsc.operator.com"},
}

// BuildSmsOipConfig builds the ApplicationConfig for SMSoIP (ap2005).
func BuildSmsOipConfig(status, provStatus, tcStatus int, configData *SmsOipConfigData) *protocol.ApplicationConfig {
	if configData == nil {
		configData = &SmsOipConfigData{}
	}

	result := &protocol.ApplicationConfig{
		AppID:             "ap2005",
		EntitlementStatus: status,
		TcStatus:          protocol.IntPtr(tcStatus),
		ProvStatus:        protocol.IntPtr(provStatus),
	}

	if status != protocol.EntitlementStatusEnabled {
		result.AddrStatus = protocol.IntPtr(0)
		return result
	}

	result.AddrStatus = protocol.IntPtr(1)
	if len(configData.Addresses) > 0 {
		result.Addresses = configData.Addresses
	} else {
		result.Addresses = _defaultSmsOipAddresses
	}

	return result
}
